qiskit/utils/deprecation.py

```diff
@@ -77,3 +92,73 @@ def _rename_kwargs(func_name, kwargs, kwarg_map, category: Type[Warning] = Depre
                 )
 
                 kwargs[new_arg] = kwargs.pop(old_arg)
+
+
+def _extend_docstring(func, version, kwarg_map):
+    """kwarg_map[None] means message in no kwarg, and it will be
+    append to the long desscription"""
+    docstr = func.__doc__
+    if docstr:
+        docstr_lines = docstr.expandtabs().splitlines()
+    else:
+        docstr_lines = ["DEPRECATED"]
+
+    indent = 1000
+    first_empty_line = None
+    for line_no, line in enumerate(docstr_lines[1:], start=1):
+        stripped = line.lstrip()
+        if stripped:
+            indent = min(indent, len(line) - len(stripped))
+        else:
+            if first_empty_line is None:
+                first_empty_line = line_no
+    if first_empty_line is None:
+        first_empty_line = len(docstr_lines)
+    spaces = ""
+    if indent != 1000:
+        spaces = " " * indent
+
+    new_doc_str_lines = docstr_lines[:first_empty_line]
+    if None in kwarg_map:
+        new_doc_str_lines += ["", spaces + f".. deprecated:: {version}"]
+        for msg_line in kwarg_map[None]:
+            new_doc_str_lines.append(spaces + "  " + msg_line)
```

Comment:
This is going to mess up if the docstring has a code block or maths block in its first paragraph, for example:

```python
r"""
This function returns the matrix:

.. math::

  \hat U = \begin{pmatrix}\ldots\end{pmatrix}
"""
```
Basically anything that introduces a Sphinx directive will have this inserted in the wrong place.

It's also not absolutely fool-proof, but perhaps a better place would be immediately before the "Args" section header if it exists, and at the end of the docstring if not?

Reply:
is it not a common thing that the first line is the "short description'? In any case, you can modify your own docstring if `docstring_version=None`. That should cover corner cases.

Reply:
It's definitely not universally followed - the Python documentation doesn't do that, for example, and I certainly wouldn't expect Terra is consistent in it. With this sort of function, people also don't manually check, so it's somewhat important for the function to not spring surprises.

Reply:
Moving the note to `Args:-1` was easier than expected. So done in def73fcf8f01ba5eb07c416f118ecc66665c0d45